File: swift/Sources/Ensoul/Resources/Frameworks.swift
/// Frameworks resource for the Ensoul Swift SDK.
///
/// Wraps all `/v1/frameworks` endpoints. Frameworks define the measurement
/// instruments (e.g. surveys, psychometric scales) used to sample personas.
///
/// Example:
/// ```swift
/// let page = try await client.frameworks.list()
/// let fw   = try await client.frameworks.get("fw_abc123")
/// let instruments = try await client.frameworks.getInstruments("fw_abc123")
/// ```
import Foundation

// MARK: - Frameworks

@available(iOS 15.0, macOS 12.0, *)
public class Frameworks {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - List

    /// GET /v1/frameworks
    ///
    /// Returns a paginated list of frameworks as raw JSON dictionaries.
    public func list(
        page: Int = 1,
        perPage: Int = 20
    ) async throws -> RawPage {
        let params: [String: String] = [
            "page": String(page),
            "per_page": String(perPage),
        ]
        let (data, _) = try await client.get("/v1/frameworks", params: params)
        return try RawPage.from(data: data)
    }

    // MARK: - Get

    /// GET /v1/frameworks/{frameworkId}
    public func get(_ frameworkId: String) async throws -> [String: Any] {
        let (data, _) = try await client.get("/v1/frameworks/\(frameworkId)")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Create

    /// POST /v1/frameworks
    ///
    /// Creates a new framework. The body shape is project-defined; pass
    /// the full configuration as a `[String: Any]` dictionary.
    public func create(_ body: [String: Any]) async throws -> [String: Any] {
        let (data, _) = try await client.post("/v1/frameworks", body: body)
        return try Self.jsonObject(from: data)
    }

    // MARK: - Update

    /// PUT /v1/frameworks/{frameworkId}
    ///
    /// Replaces the framework configuration. Provide the full updated body.
    public func update(_ frameworkId: String, body: [String: Any]) async throws -> [String: Any] {
        let (data, _) = try await client.put("/v1/frameworks/\(frameworkId)", body: body)
        return try Self.jsonObject(from: data)
    }

    // MARK: - Delete

    /// DELETE /v1/frameworks/{frameworkId}
    public func delete(_ frameworkId: String) async throws {
        _ = try await client.delete("/v1/frameworks/\(frameworkId)")
    }

    // MARK: - Validations

    /// GET /v1/frameworks/{frameworkId}/validations
    ///
    /// Returns the framework's validation results
    /// (e.g. `{ "valid": true, "errors": [] }`).
    public func validations(_ frameworkId: String) async throws -> [String: Any] {
        let (data, _) = try await client.get(
            "/v1/frameworks/\(frameworkId)/validations"
        )
        return try Self.jsonObject(from: data)
    }

    // MARK: - Get Instruments

    /// GET /v1/frameworks/{frameworkId}/instruments
    ///
    /// Returns all measurement instruments belonging to this framework as raw
    /// JSON dictionaries (instrument schema is project-specific).
    public func getInstruments(_ frameworkId: String) async throws -> [[String: Any]] {
        let (data, _) = try await client.get("/v1/frameworks/\(frameworkId)/instruments")
        // The server may return a bare array or a wrapped object.
        if let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
            return array
        }
        if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let items = dict["items"] as? [[String: Any]] {
            return items
        }
        throw EnsoulAPIError(
            statusCode: 200,
            error: "ParseError",
            message: "Expected a JSON array or paginated object for framework instruments"
        )
    }

    // MARK: - Private helpers

    private static func jsonObject(from data: Data) throws -> [String: Any] {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EnsoulAPIError(
                statusCode: 200,
                error: "ParseError",
                message: "Expected a JSON object in framework response"
            )
        }
        return dict
    }
}

File: kotlin/src/main/kotlin/ai/ensoul/sdk/resources/Frameworks.kt
package ai.ensoul.sdk.resources

import ai.ensoul.sdk.EnsoulHttpClient
import ai.ensoul.sdk.Page
import io.ktor.client.statement.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class Frameworks(private val client: EnsoulHttpClient) {

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun list(page: Int = 1, perPage: Int = 20): Page<JsonObject> {
        val params = mapOf<String, Any?>("page" to page, "per_page" to perPage)
        val response = client.get("/v1/frameworks", params = params)
        val text = response.bodyAsText()
        val data = json.parseToJsonElement(text).jsonObject
        val rawItems = data["items"]?.jsonArray?.map { it.jsonObject } ?: emptyList()
        return Page(
            items = rawItems,
            total = data["total"]?.jsonPrimitive?.intOrNull ?: rawItems.size,
            page = data["page"]?.jsonPrimitive?.intOrNull ?: page,
            perPage = data["per_page"]?.jsonPrimitive?.intOrNull ?: perPage,
            pages = data["pages"]?.jsonPrimitive?.intOrNull ?: 1,
            client = client,
            method = "GET",
            path = "/v1/frameworks",
            params = params,
            deserializer = { it },
        )
    }

    suspend fun get(frameworkId: String): JsonObject {
        val response = client.get("/v1/frameworks/$frameworkId")
        return json.parseToJsonElement(response.bodyAsText()).jsonObject
    }

    suspend fun create(body: Map<String, Any?>): JsonObject {
        val response = client.post("/v1/frameworks", json = body)
        return json.parseToJsonElement(response.bodyAsText()).jsonObject
    }

    suspend fun update(frameworkId: String, body: Map<String, Any?>): JsonObject {
        val response = client.put("/v1/frameworks/$frameworkId", json = body)
        return json.parseToJsonElement(response.bodyAsText()).jsonObject
    }

    suspend fun delete(frameworkId: String) {
        client.delete("/v1/frameworks/$frameworkId")
    }

    /** GET /v1/frameworks/{frameworkId}/validations */
    suspend fun validations(frameworkId: String): JsonObject {
        val response = client.get("/v1/frameworks/$frameworkId/validations")
        return json.parseToJsonElement(response.bodyAsText()).jsonObject
    }

    suspend fun getInstruments(frameworkId: String): List<JsonObject> {
        val response = client.get("/v1/frameworks/$frameworkId/instruments")
        val data = json.parseToJsonElement(response.bodyAsText())
        return if (data is JsonArray) {
            data.map { it.jsonObject }
        } else {
            data.jsonObject["items"]?.jsonArray?.map { it.jsonObject } ?: emptyList()
        }
    }
}
